File: MoveTrack-iOS/widowmaker/MoveTrack/Services/ClientEventLogger.swift
//
//  ClientEventLogger.swift
//  Nexus Moves
//
//  Lightweight batched client-event log for the scan/chat pipeline. Before
//  this, a client-side failure (an aborted upload, an SSE timeout, a review
//  card discarded on error) left no trace anywhere — see
//  docs/notes/beta-scan-reliability-investigation.md Section 5.3 #10.
//
//  This is NOT a crash reporter. It intentionally does not capture stack
//  traces, device identifiers beyond OS/app version, or arbitrary PII, and it
//  does not retry failed sends — it's a best-effort lifecycle log, not a
//  delivery guarantee. Full crash/error reporting (Sentry, Crashlytics, or
//  similar) is a real gap this doesn't close; that's a vendor decision left
//  for Owen to make deliberately rather than picked implicitly here.
//

import Foundation
#if canImport(UIKit)
import UIKit
#endif

enum ClientEventType: String, Encodable {
    case uploadStarted = "upload_started"
    case uploadSucceeded = "upload_succeeded"
    case uploadFailed = "upload_failed"
    case sseTimeout = "sse_timeout"
    case turnFailed = "turn_failed"
    case reviewCardShown = "review_card_shown"
    case reviewCardCommitted = "review_card_committed"
}

/// An `actor` — not a plain class — because `log()` is called from many
/// contexts at once (including synchronous, non-async call sites like
/// `NexusViewModel.handle(_:)`) while `flush()` runs on a timer. A plain class
/// let `log()`'s buffer append (implicitly MainActor, since all callers today
/// are on MainActor) race against `flush()`'s unstructured `Task {}`, which
/// does NOT inherit MainActor isolation — lost/duplicated events, and it
/// wouldn't compile under Swift 6 strict concurrency at all.
actor ClientEventLogger {
    static let shared = ClientEventLogger()

    private struct Event: Encodable {
        let type: String
        let sessionId: String?
        let clientAt: String
        let metadata: [String: String]?
    }

    private struct Batch: Encodable {
        let events: [Event]
        let appVersion: String?
        let osVersion: String?
    }

    private let encoder = JSONEncoder()
    private var buffer: [Event] = []
    private var flushTask: Task<Void, Never>?

    private let maxBufferSize = 20
    private let flushDelayNanoseconds: UInt64 = 10_000_000_000 // 10s

    private init() {}

    /// Queue an event for the next batch. `nonisolated` so every existing call
    /// site keeps working unawaited from any context (sync or async); the
    /// actual buffer mutation hops onto the actor via `Task`, so it's still
    /// race-free — just not synchronous with the call.
    nonisolated func log(_ type: ClientEventType, sessionId: String? = nil, metadata: [String: String]? = nil) {
        let event = Event(
            type: type.rawValue,
            sessionId: sessionId,
            clientAt: ISO8601DateFormatter().string(from: Date()),
            metadata: metadata
        )
        Task { await self.enqueue(event) }
    }

    private func enqueue(_ event: Event) {
        buffer.append(event)
        if buffer.count >= maxBufferSize {
            flushTask?.cancel()
            flushTask = Task { await flush() }
        } else {
            scheduleFlush()
        }
    }

    private func scheduleFlush() {
        guard flushTask == nil else { return }
        flushTask = Task { [weak self, flushDelayNanoseconds] in
            try? await Task.sleep(nanoseconds: flushDelayNanoseconds)
            await self?.flush()
        }
    }

    /// Best-effort, fire-and-forget — a failed send drops the batch rather than
    /// retrying or blocking the caller. Call directly (e.g. on scene
    /// backgrounding) to flush without waiting for the timer.
    @discardableResult
    func flush() async -> Bool {
        flushTask = nil
        guard !buffer.isEmpty else { return true }
        guard let token = AuthService.shared.sessionToken,
              let url = URL(string: Constants.apiBaseURL + "/api/client-events") else {
            buffer.removeAll()
            return false
        }

        let toSend = buffer
        buffer = []

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let appVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String
        #if canImport(UIKit)
        let osVersion: String? = UIDevice.current.systemVersion
        #else
        let osVersion: String? = nil
        #endif

        do {
            request.httpBody = try encoder.encode(Batch(events: toSend, appVersion: appVersion, osVersion: osVersion))
            _ = try await URLSession.shared.data(for: request)
            return true
        } catch {
            return false
        }
    }
}

File: MoveTrack-Android/app/src/main/java/dev/we3kings/nexusmoves/data/events/ClientEventLogger.kt
package dev.we3kings.nexusmoves.data.events

import android.os.Build
import dev.we3kings.nexusmoves.BuildConfig
import dev.we3kings.nexusmoves.data.Constants
import dev.we3kings.nexusmoves.data.auth.TokenStore
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.time.Instant
import java.util.concurrent.TimeUnit

/**
 * Lightweight batched client-event log for the scan/chat pipeline — port of iOS
 * ClientEventLogger. Same event names so dashboards stay unified. NOT a crash
 * reporter: no stack traces, no PII, no retries — best-effort lifecycle only.
 */
object ClientEventLogger {

    enum class Type(val raw: String) {
        UploadStarted("upload_started"),
        UploadSucceeded("upload_succeeded"),
        UploadFailed("upload_failed"),
        SseTimeout("sse_timeout"),
        TurnFailed("turn_failed"),
        ReviewCardShown("review_card_shown"),
        ReviewCardCommitted("review_card_committed"),
    }

    @Serializable
    private data class Event(
        val type: String,
        val sessionId: String? = null,
        val clientAt: String,
        val metadata: Map<String, String>? = null,
    )

    @Serializable
    private data class Batch(val events: List<Event>, val appVersion: String?, val osVersion: String?)

    private val json = Json { encodeDefaults = true; explicitNulls = false }
    private val jsonMedia = "application/json".toMediaType()
    private val http = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .build()

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val mutex = Mutex()
    private val buffer = mutableListOf<Event>()
    private var flushJob: Job? = null

    private const val MAX_BUFFER = 20
    private const val FLUSH_DELAY_MS = 10_000L

    /** Queue an event for the next batch. Fire-and-forget from any thread. */
    fun log(type: Type, sessionId: String? = null, metadata: Map<String, String>? = null) {
        val event = Event(type.raw, sessionId, Instant.now().toString(), metadata)
        scope.launch { enqueue(event) }
    }

    private suspend fun enqueue(event: Event) {
        val shouldFlushNow = mutex.withLock {
            buffer.add(event)
            buffer.size >= MAX_BUFFER
        }
        if (shouldFlushNow) {
            flushJob?.cancel()
            flushJob = scope.launch { flush() }
        } else {
            scheduleFlush()
        }
    }

    private fun scheduleFlush() {
        if (flushJob?.isActive == true) return
        flushJob = scope.launch {
            delay(FLUSH_DELAY_MS)
            flush()
        }
    }

    /** Flush the buffer now (e.g. on process background). Best-effort. */
    suspend fun flush() {
        val toSend = mutex.withLock {
            flushJob = null
            if (buffer.isEmpty()) return
            val copy = buffer.toList()
            buffer.clear()
            copy
        }
        val token = TokenStore.sessionToken ?: return
        val batch = Batch(
            events = toSend,
            appVersion = BuildConfig.VERSION_NAME,
            osVersion = Build.VERSION.RELEASE,
        )
        try {
            val request = Request.Builder()
                .url(Constants.apiBaseURL + "/api/client-events")
                .header("Content-Type", "application/json")
                .header("Authorization", "Bearer $token")
                .post(json.encodeToString(batch).toRequestBody(jsonMedia))
                .build()
            http.newCall(request).execute().use { /* drop the response */ }
        } catch (_: Exception) {
            // Best-effort — drop on failure (iOS parity).
        }
    }

    /** Fire a flush without awaiting (scene-background hook). */
    fun flushBlocking() {
        scope.launch { flush() }
    }
}
